core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -0,0 +1,124 @@
+}
+
+@SuppressAnimalSniffer
+private class ClassValueCache(private val compute: (KClass<*>) -> KSerializer<Any>?) : SerializerCache {
+    private val classValue = initClassValue()
+
+    private fun initClassValue() = object : ClassValue<CacheEntry>() {
+        override fun computeValue(type: Class<*>): CacheEntry {
+            val pair = SerializerPair.from(compute(type.kotlin))
+            return CacheEntry(pair)
+        }
+    }
+
+    override fun get(key: KClass<Any>, isNullable: Boolean): KSerializer<Any?>? =
+        classValue[key.java].serializers?.get(isNullable)
+}
+
+@SuppressAnimalSniffer
+private class ClassValueParametrizedCache(private val compute: (KClass<Any>, List<KType>) -> KSerializer<Any>?) : ParametrizedSerializerCache {
+    private val classValue = initClassValue()
+
+    private fun initClassValue() = object : ClassValue<ParametrizedCacheEntry>() {
+        override fun computeValue(type: Class<*>): ParametrizedCacheEntry {
+            return ParametrizedCacheEntry()
+        }
+    }
+
+    override fun get(key: KClass<Any>, isNullable: Boolean, types: List<KType>): Result<KSerializer<Any?>?> =
+        classValue[key.java]
+            .computeIfAbsent(types) { compute(key, types) }
+            .map { it?.get(isNullable) }
+}
+
+/**
+ * We no longer support Java 6, so the only place we use this cache is Android, where there
+ * are no classloader leaks issue, thus we can safely use strong references and do not bother
+ * with WeakReference wrapping.
+ */
+private class ConcurrentHashMapCache(private val compute: (KClass<*>) -> KSerializer<Any>?) : SerializerCache {
+    private val cache = ConcurrentHashMap<Class<*>, CacheEntry>()
+
+    override fun get(key: KClass<Any>, isNullable: Boolean): KSerializer<Any?>? {
+        return cache.getOrPut(key.java) {
+            CacheEntry(SerializerPair.from(compute(key)))
+        }.serializers?.get(isNullable)
+    }
+}
+
+
+
+private class ConcurrentHashMapParametrizedCache(private val compute: (KClass<Any>, List<KType>) -> KSerializer<Any>?) : ParametrizedSerializerCache {
+    private val cache = ConcurrentHashMap<Class<*>, ParametrizedCacheEntry>()
+
+    override fun get(key: KClass<Any>, isNullable: Boolean, types: List<KType>): Result<KSerializer<Any?>?> {
+        return cache.getOrPut(key.java) { ParametrizedCacheEntry() }
+            .computeIfAbsent(types) { compute(key, types) }
+            .map { it?.get(isNullable) }
+    }
+}
+
+
+
+private class SerializerPair(private val nonNull: KSerializer<out Any>, private val nullable: KSerializer<out Any?>) {
+    fun get(isNullable: Boolean): KSerializer<Any?> {
+        return (if (isNullable) nullable else nonNull).cast()
+    }
+
+    companion object {
+        fun from(serializer: KSerializer<out Any>?): SerializerPair? {
+            return serializer?.let { SerializerPair(it, it.nullable) }
```

Comment:
I'm not sure creating a pair of serializers right away is a great idea. Since this is a cache only for top-level serializers, I strongly feel that many nullable serializers will never be used, since the non-nullable case is much, much more often. In the end, they'll be just hard-referenced garbage in memory

Reply:
I second that, that's a good observation.
A separate cache for nullable serializer might be good